### DISCUSSION TITLE
Refine Zod schema for MySQL types

### DISCUSSION THREAD

**Discussion by TonyGravagno** [araera111/mysql-to-zod#21] [Ideas]
I have a lot of code now for Zod schema that is much more rigorous than the z.types returned from toZod.ts. For example, for [DATE](https://github.com/araera111/mysql-to-zod/blob/bc18e8cd1fa1e5c35e5425264c9ac0ac3ba63ddb/src/toZod.spec.ts#L15C1-L15C1), DATETIME, TIMESTAMP, and YEAR I have extensive schema that can be implemented as follows:

The existing `return z.type()` can be replaced with `return schemaType`. The types are extensive and should not be duplicated for every table. For dates, for example:

```ts
export const schemaDate: z.ZodType<Date, any, string> = z
  .string()
  .refine(
    value => {
...
    },
    { message: 'Invalid DATE' }
  )
  .transform(getDateObject)

export const schemaDateTime: z.ZodType<Date, any, string> = z
...
export const schemaTimeStamp: z.ZodType<Date, any, string> = z
...
export const schemaYear: z.ZodType<number | string, any, string | number> = z
...

    case "DATE":
       return 'schemaDate'
    case "DATETIME":
       return 'schemaDateTime'
    case "TIMESTAMP":
       return 'schemaTimeStamp'
    case "YEAR":
       return 'schemaYear'
```

The reason for this extensive checking is that MySQL has special rules about types, which are not checked with simple Zod schema. I have been coding to specs in the MySQL v8 reference, for [DATE types](https://dev.mysql.com/doc/refman/8.0/en/datetime.html) and [YEAR](https://dev.mysql.com/doc/refman/8.0/en/year.html).

I am testing now, separate from MySQL-to-Zod. If you agree to implement PRs with these enhancements I will contribute after integration. I need to check the upstream Zod code to see if I'm duplicating functionality that already exists.

I believe there should be a config option so that a developer can use the simple defaults, like z.date(), or quickly switch use the more extensive version to see if that works better.

If you agree that a fall-back capability will be good in options, there are a few more options I'd like to consider. For example, there are MySQL server settings that change the rules of types. It would help if these new extensive validations could be tuned so that they agree with server settings. This is a longer-term effort. Also, MySQL v8 is different from prior versions - the validation code should adapt to the server with flags. I'm also thinking that there should be generation-time flags that determine the code that is generated, **and** runtime flags, so that when a single schema is invoked at runtime it behaves correctly for the target database server.

What do you think of the general concepts so far?

Thanks!

**Comment by araera111** [comment]
I would like to improve this library and am very happy with your suggestions. I feel very pleased to have an opportunity to think about it. I am not at all opposed to adding options. Just for the record.

I agree with you about refining zodSchema for MySQL types.

I think it would be good to create an option and add custom validation logic more in line with MySQL limitations in the refinement.
I would be willing to add this to the options if you provide the source code.

My question is how is this zodSchema used?

I primarily use the schema when connecting to the DB using knex. I use it as follows.

```ts
const testSchema = z.object({
  id: z.number(),
  name: z.string(),
});

type Test = z.infer<typeof testSchema>;
const result = await db<Test>("test").select("*");
```
Year is taken as number, Date, DateTime, TimeStamp as Date type.
This is the default value of mysql2.
If you pass the option ```dateStrings```, you can get them as string type, but I am still considering whether to support this or not.

Is your refine (custom validation) used before you insert or update data into mysql?
Is it there to check the type first and see if it is ok to INSERT it?
Or is it used at another time?

I think the max year for the TimeStamp type is up to 2038, but if it is more than that, it will give an error?
That way I can notice the error before inserting it. Is it used in such cases?

It doesn't seem like it would make much sense to apply strict validation and get an error when selecting data with knex, etc. instead of insert or update.

I think it might be useful and could be added, but I don't know the use case.
My apologies for my lack of imagination.

Sorry for the long sentence.

**Comment by TonyGravagno** [comment]
As coded now the schema is intended to validate compliance with MySQL for insert/update. So, for my purposes at this time, with a form on a page, if a user enters a date like 1400-01-01, the schema will not validate and the UI will show an error. Also, if the user enters 2023-12-33, the Date class does validate this to 2024-01-02, but the new extended schema catches that.

If the data comes from another source, like a CSV, the date might have two digit years. MySQL has specific rules about which century gets applied to a specific two digit "window" : (I think) 00-69 = 2000 and 70-99 = 1900. For a given application this might not be desirable. With the new `schemaYear` schema, MySQL-to-Zod has a single point in the application to set the constraint for data input. 

For current business, it is not unusual for a 30 year agreement now to terminate in 2053. So modern databases Must be configured to allow a date maximum other than the default. By putting a new maximum in MySQL-to-Zod and schema options, the `schemaDate` and other schema validations will be generated to agree with the server.

Similarly, I don't know how MySQL is tuned for historical data, like family ancestors, museums, etc, but years from 0 to 1900 may be valid and the schema should have an option to allow for this.

So there are two kinds of "options": 1) MySQL-to-Zod options to generate specific kinds of validation. 2) Runtime options so that the schema can be toggled with application-specific limits without changing the generated code.

I am happy to answer any other questions!! And VERY happy to contribute to your fine project.

Don't worry about long exchanges, this is expected for new initiatives and introductions - and I tend to write a lot. 😉

**Comment by araera111** [comment]
I understand your suggestion.
Indeed, it would be beneficial to create the data on the front end before doing insert, update.
We will follow your suggestion and make improvements.

One thing to consider is perhaps the gradual increase in complexity of options.
I believe this is an unavoidable part of creating a useful tool.
If you are a loose type and want to use it easily, you can use it without config.

There are also cases where you may want to change the strictness of the type when SELECT, (INSERT, UPDATE).
Data that should not be included in the DB may be included due to loose DB settings.
In this case, you may want to use a loose format for SELECT, but strictly enforce it when INSERT or UPDATE in the future.

```ts
export const SELECTTodoSchema = z.object({
  id: z.number(),
  task: z.string(),
  due_date: z.date().nullish(),
  created_at: z.date().nullish(),
  updated_at: z.date().nullish(),
  comment: z.string().nullish(),
});

export const UPDATEINSERTTodoSchema = z.object({
  id: z.number().nullish(), // when update, null is ok. 
  task: z.string(),
  due_date: z.date().nullish(), // more strict date type.
  created_at: z.date().nullish(),
  updated_at: z.date().nullish(),
  comment: z.string().nullish(),
});
```

We need to consider prefixes, but perhaps we could consider generating such a type.
But this is another suggestion.
Sorry for going off topic.

I agree with you and would like to make it more finely configurable.
Thanks for your input and for sharing the source code.

**Comment by TonyGravagno** [comment]
Since I already have a lot of code for date types, may I contribute my code for your review after I have done more thorough testing? All tests are passing but I have not yes incorporate my code into an actual application yet to test how it really works. If I just provide the code that generates the schema, and you approve, you or I can incorporate it into this package.

Since this is still in live development it might take a few weeks for me to submit a final proposal.

**Comment by araera111** [comment]
Thanks for your suggestion.
It would be very helpful if you could share the date type code with me.
